### ISSUE TITLE
[Bug/Feature] Pixel 2 XL Not Scanning When QR Code is Exact Fit in ViewFinder

### ISSUE THREAD

**Issue by hkmushtaq** [yuriy-budiyev/code-scanner#10]
The issue I have seen on the Pixel 2 XL and I believe the Pixel 2 as well is that for some reason if the QR Code is an exact fit in the view finder frame, it does not scan. I have to move the phone back a bit a relatively considerable amount (QR Code smaller in camera ~8-10 inches) and then it scans properly.

And in my testing (testing various devices, Samsung, LG, Pixel 1, Pixel 2) the Pixel 2/XL are the only phones to have this issue.

**Why is this a problem?**
Users will habitually try to fit the QR Code in the frame and not think to move it back because its counter intuitive

**Proposed Solution**
Add option to the code scanner to pass in the full width rect instead of the rect of the viewfinder (currently fixed at 0.75 of width and with my PR that is adjustable) to the decoder task so that it scans a larger area than the view finder and is able to scan the QR Code on devices that may have this issue

Obviously this would be an option with perhaps the following api
codeScanner.setScanFrameOverride(true|false)
codeScanner.setScanFrameSize(0.1 <= frameSize <= 1.0)

or even
codeScanner.setFullWidthScan(true|false) which would tell the codeScanner to use the full width frame of the viewfinderview instead of frameRect(which it is doing right now)

**Comment by yuriy-budiyev** [comment]
Hello.
As I think, what is really counterintuitive - if outside of frame rect will be recognized.
Btw, I see another soulution, make real recognition area a bit larger than viewfinder frame.
I'll implement something like this in next version.

**Comment by yuriy-budiyev** [comment]
Seems like the problem isn't in frame or scanning area. It is in ZXing library itself. Because when you scan from small distance from screen, phone's camera sees different subpixels (RGB) and the library can't decode such images well.

**Comment by hkmushtaq** [comment]
> Hello.
> As I think, what is really counterintuitive - if code outside of frame rect will be recognized.
> Btw, I see another soulution, make real recognition area a bit larger than viewfinder frame.
> I'll implement something like this in next version.

The solution you mentioned is what I was proposing in the original post albeit it could have a different approach to do it. Apologies if I wasn't clear enough on that. I am willing to discuss and do the implementation if you agree with the solution we arrive at

**Comment by yuriy-budiyev** [comment]
I've already implemented this, and it got no result even with 20% larger area. See my last comment. I'll get Pxel 2 XL in a couple of weeks and test library.
Anyways, completely ignore the frame is bad idea.

**Comment by hkmushtaq** [comment]
Just some more info

I have not been able to reproduce this issue on Samsung Galaxy S8 which I believe also has the same screen aspect ratio. The issue seems to just be for Pixel 2 and Pixel 2 XL Devices.

**Comment by yuriy-budiyev** [comment]
So, this problem should be investigated deeper. I'll kep this issue open.
What parameters of scanner view are you using? (full screen or not, viewfinder aspect ration, etc)

**Comment by hkmushtaq** [comment]
Just off the top of my head, do you think it would improve the reliability of the decoder if we did a high contrast filter on the frame before decoding it?

**Comment by hkmushtaq** [comment]
I was able to resolve this issue by reapplying your threshold changes and also utilizing a smaller frameSize (was added in my previous PR). Now it scans perfectly for me. I've made it so that the threshold is configurable and is 0 by default (existing behavior) and opened a PR so you can take a look at the code

**Comment by hkmushtaq** [comment]
So I investigated this issue some more. I thought I had resolved the issue with the changes mentioned above but then on a different monitor, it did not work (most likely due to the subpixels issue you mentioned briefly above)

The next thing I tried was changing the LuminanceSource from PlanarYUV to RGB and that actually ended up resolving that issue. I will test on various devices and monitors/screens to ensure it works consistently and then make a PR.

One thing to note. I think that not all devices actually support RGB mode for the Camera so I am currently converting the cropped frame ONLY to RGB and creating a RGBLuminanceSource
